File: C++/Interview_problems/NotificationSystem_demo.cpp
/*

Functional Requirements:
    User should be able to send notifiation
    Notification system should support multiple types of channels
    notification system will support user preferences
    NS should process the notification asynchronous
    Retry failed notifications

Non-Functional Requirement:
    Error handling
    Asynchronous
    Atleast once delivery


Notification
NotificationService
NotificationDispather
NotificationChannel
    SMSNotificationChannel
    WhatsappNotificationChannel
    .....
NotificationFactory
User

*/

#include <atomic>
#include <chrono>
#include <condition_variable>
#include <iostream>
#include <memory>
#include <mutex>
#include <queue>
#include <set>
#include <sstream>
#include <string>
#include <thread>
#include <unordered_map>
#include <vector>

enum class NotificationType { SMS, WHATSAPP, EMAIL };
enum class NotificationStatus { PENDING, SENT, FAILED };

struct User {
    std::string id;
    std::string name;
};

struct Notification {
    std::string id;
    std::shared_ptr<User> user;
    std::string recipientId;
    std::string message;
    int priority = 0; // lower value = higher priority
    NotificationType type;
    NotificationStatus status = NotificationStatus::PENDING;
};

struct NotificationChannel {
    virtual ~NotificationChannel() = default;
    virtual bool sendNotification(const std::shared_ptr<Notification>& notification) = 0;
};

class SMSNotificationChannel : public NotificationChannel {
public:
    bool sendNotification(const std::shared_ptr<Notification>& notification) override {
        std::cout << "[SMS] to " << notification->recipientId << ": " << notification->message << "\n";
        return true;
    }
};

class WhatsappNotificationChannel : public NotificationChannel {
public:
    bool sendNotification(const std::shared_ptr<Notification>& notification) override {
        std::cout << "[WhatsApp] to " << notification->recipientId << ": " << notification->message << "\n";
        return true;
    }
};

class EmailNotificationChannel : public NotificationChannel {
public:
    bool sendNotification(const std::shared_ptr<Notification>& notification) override {
        std::cout << "[Email] to " << notification->recipientId << ": " << notification->message << "\n";
        return true;
    }
};

class NotificationChannelFactory {
public:
    NotificationChannelFactory() {
        registry_[NotificationType::SMS] = std::make_shared<SMSNotificationChannel>();
        registry_[NotificationType::WHATSAPP] = std::make_shared<WhatsappNotificationChannel>();
        registry_[NotificationType::EMAIL] = std::make_shared<EmailNotificationChannel>();
    }

    std::shared_ptr<NotificationChannel> getNotificationChannel(NotificationType type) {
        auto it = registry_.find(type);
        if (it == registry_.end()) {
            throw std::invalid_argument("No channel registered for type");
        }
        return it->second;
    }

private:
    std::unordered_map<NotificationType, std::shared_ptr<NotificationChannel>> registry_;
};

class UserPreferenceService {
public:
    void setPreferences(const std::string& userId, std::set<NotificationType> types) {
        std::lock_guard<std::mutex> lock(mutex_);
        preferences_[userId] = std::move(types);
    }

    std::set<NotificationType> getPreferences(const std::string& userId) {
        std::lock_guard<std::mutex> lock(mutex_);
        auto it = preferences_.find(userId);
        if (it == preferences_.end()) {
            return {};
        }
        return it->second;
    }

private:
    std::mutex mutex_;
    std::unordered_map<std::string, std::set<NotificationType>> preferences_;
};

// Thread-safe priority queue keyed on Notification::priority (lower = more urgent).
// A null shared_ptr acts as the poison pill used to shut the worker down.
class NotificationRequestQueue {
public:
    void offer(std::shared_ptr<Notification> notification) {
        std::lock_guard<std::mutex> lock(mutex_);
        int priority = notification ? notification->priority : -1;
        heap_.push({priority, seq_++, std::move(notification)});
        cv_.notify_one();
    }

    std::shared_ptr<Notification> take() {
        std::unique_lock<std::mutex> lock(mutex_);
        cv_.wait(lock, [this] { return !heap_.empty(); });
        auto top = heap_.top();
        heap_.pop();
        return top.notification;
    }

private:
    struct Entry {
        int priority;
        long seq;
        std::shared_ptr<Notification> notification;

        bool operator<(const Entry& other) const {
            if (priority != other.priority) return priority > other.priority; // min-heap on priority
            return seq > other.seq;
        }
    };

    std::mutex mutex_;
    std::condition_variable cv_;
    std::priority_queue<Entry> heap_;
    long seq_ = 0;
};

class NotificationDispatcher {
public:
    void enqueue(std::shared_ptr<Notification> notification) { queue_.offer(std::move(notification)); }
    std::shared_ptr<Notification> nextTask() { return queue_.take(); }

private:
    NotificationRequestQueue queue_;
};

class NotificationWorker {
public:
    static constexpr int MAX_RETRIES = 3;

    NotificationWorker(NotificationDispatcher& dispatcher,
                        UserPreferenceService& preferenceService,
                        NotificationChannelFactory& channelFactory)
        : dispatcher_(dispatcher), preferenceService_(preferenceService), channelFactory_(channelFactory) {}

    void run() {
        while (true) {
            auto notification = dispatcher_.nextTask();
            if (!notification) break; // poison pill

            auto userPref = preferenceService_.getPreferences(notification->recipientId);
            std::set<NotificationType> channelsToUse = userPref.empty()
                    ? std::set<NotificationType>{notification->type}
                    : userPref;

            std::vector<std::thread> deliveries;
            for (auto type : channelsToUse) {
                deliveries.emplace_back(&NotificationWorker::deliverWithRetry, this, notification, type);
            }
            for (auto& t : deliveries) t.join();
        }
    }

private:
    void deliverWithRetry(std::shared_ptr<Notification> notification, NotificationType type) {
        auto channel = channelFactory_.getNotificationChannel(type);
        for (int attempt = 1; attempt <= MAX_RETRIES; ++attempt) {
            try {
                if (channel->sendNotification(notification)) {
                    notification->status = NotificationStatus::SENT;
                    return;
                }
            } catch (const std::exception& e) {
                std::cout << "Attempt " << attempt << " failed: " << e.what() << "\n";
            }
        }
        notification->status = NotificationStatus::FAILED;
        std::cout << "Notification " << notification->id << " failed after " << MAX_RETRIES << " attempts\n";
    }

    NotificationDispatcher& dispatcher_;
    UserPreferenceService& preferenceService_;
    NotificationChannelFactory& channelFactory_;
};

class NotificationService {
public:
    explicit NotificationService(NotificationDispatcher& dispatcher) : dispatcher_(dispatcher) {}

    bool submitNotificationRequest(std::shared_ptr<Notification> notification) {
        dispatcher_.enqueue(std::move(notification));
        return true;
    }

private:
    NotificationDispatcher& dispatcher_;
};

int main() {
    UserPreferenceService preferenceService;
    NotificationChannelFactory channelFactory;
    NotificationDispatcher dispatcher;
    NotificationService notificationService(dispatcher);

    auto alice = std::make_shared<User>(User{"u1", "Alice"});
    preferenceService.setPreferences("u1", {NotificationType::EMAIL, NotificationType::SMS});

    NotificationWorker worker(dispatcher, preferenceService, channelFactory);
    std::thread workerThread(&NotificationWorker::run, &worker);

    auto n1 = std::make_shared<Notification>();
    n1->id = "n1";
    n1->user = alice;
    n1->recipientId = "u1";
    n1->message = "Your order has shipped!";
    n1->priority = 1;
    n1->type = NotificationType::EMAIL;
    notificationService.submitNotificationRequest(n1);

    auto n2 = std::make_shared<Notification>();
    n2->id = "n2";
    n2->user = alice;
    n2->recipientId = "u1";
    n2->message = "OTP: 4821";
    n2->priority = 0;
    n2->type = NotificationType::SMS;
    notificationService.submitNotificationRequest(n2);

    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    dispatcher.enqueue(nullptr); // poison pill to stop the worker
    workerThread.join();

    std::cout << "Done.\n";
    return 0;
}

File: Golang/Interview_problems/NotificationSystem_demo.go
/*
Functional Requirements:
    User should be able to send notifiation
    Notification system should support multiple types of channels
    notification system will support user preferences
    NS should process the notification asynchronous
    Retry failed notifications

Non-Functional Requirement:
    Error handling
    Asynchronous
    Atleast once delivery


Notification
NotificationService
NotificationDispather
NotificationChannel
    SMSNotificationChannel
    WhatsappNotificationChannel
    .....
NotificationFactory
User
*/

package main

import (
	"container/heap"
	"fmt"
	"sync"
	"time"
)

type NotificationType int

const (
	SMS NotificationType = iota
	WHATSAPP
	EMAIL
)

type NotificationStatus int

const (
	PENDING NotificationStatus = iota
	SENT
	FAILED
)

type User struct {
	ID   string
	Name string
}

type Notification struct {
	ID          string
	User        *User
	RecipientID string
	Message     string
	Priority    int // lower value = higher priority
	Type        NotificationType
	Status      NotificationStatus
}

type NotificationChannel interface {
	SendNotification(notification *Notification) bool
}

type SMSNotificationChannel struct{}

func (c *SMSNotificationChannel) SendNotification(notification *Notification) bool {
	fmt.Printf("[SMS] to %s: %s\n", notification.RecipientID, notification.Message)
	return true
}

type WhatsappNotificationChannel struct{}

func (c *WhatsappNotificationChannel) SendNotification(notification *Notification) bool {
	fmt.Printf("[WhatsApp] to %s: %s\n", notification.RecipientID, notification.Message)
	return true
}

type EmailNotificationChannel struct{}

func (c *EmailNotificationChannel) SendNotification(notification *Notification) bool {
	fmt.Printf("[Email] to %s: %s\n", notification.RecipientID, notification.Message)
	return true
}

type NotificationChannelFactory struct {
	registry map[NotificationType]NotificationChannel
}

func NewNotificationChannelFactory() *NotificationChannelFactory {
	return &NotificationChannelFactory{
		registry: map[NotificationType]NotificationChannel{
			SMS:      &SMSNotificationChannel{},
			WHATSAPP: &WhatsappNotificationChannel{},
			EMAIL:    &EmailNotificationChannel{},
		},
	}
}

func (f *NotificationChannelFactory) GetNotificationChannel(t NotificationType) (NotificationChannel, error) {
	channel, ok := f.registry[t]
	if !ok {
		return nil, fmt.Errorf("no channel registered for type: %v", t)
	}
	return channel, nil
}

type UserPreferenceService struct {
	mu          sync.Mutex
	preferences map[string]map[NotificationType]bool
}

func NewUserPreferenceService() *UserPreferenceService {
	return &UserPreferenceService{preferences: make(map[string]map[NotificationType]bool)}
}

func (s *UserPreferenceService) SetPreferences(userID string, types []NotificationType) {
	s.mu.Lock()
	defer s.mu.Unlock()
	set := make(map[NotificationType]bool, len(types))
	for _, t := range types {
		set[t] = true
	}
	s.preferences[userID] = set
}

func (s *UserPreferenceService) GetPreferences(userID string) map[NotificationType]bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.preferences[userID]
}

// entry wraps a notification (nil = poison pill) for the priority queue.
type entry struct {
	priority     int
	seq          int
	notification *Notification
}

type entryHeap []*entry

func (h entryHeap) Len() int { return len(h) }
func (h entryHeap) Less(i, j int) bool {
	if h[i].priority != h[j].priority {
		return h[i].priority < h[j].priority
	}
	return h[i].seq < h[j].seq
}
func (h entryHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *entryHeap) Push(x interface{}) { *h = append(*h, x.(*entry)) }
func (h *entryHeap) Pop() interface{} {
	old := *h
	n := len(old)
	item := old[n-1]
	*h = old[:n-1]
	return item
}

// NotificationRequestQueue is a thread-safe priority queue keyed on Notification.Priority
// (lower = more urgent). A nil notification acts as the poison pill for shutdown.
type NotificationRequestQueue struct {
	mu   sync.Mutex
	cond *sync.Cond
	heap entryHeap
	seq  int
}

func NewNotificationRequestQueue() *NotificationRequestQueue {
	q := &NotificationRequestQueue{}
	q.cond = sync.NewCond(&q.mu)
	return q
}

func (q *NotificationRequestQueue) Offer(notification *Notification) {
	q.mu.Lock()
	priority := -1
	if notification != nil {
		priority = notification.Priority
	}
	heap.Push(&q.heap, &entry{priority: priority, seq: q.seq, notification: notification})
	q.seq++
	q.mu.Unlock()
	q.cond.Signal()
}

func (q *NotificationRequestQueue) Take() *Notification {
	q.mu.Lock()
	defer q.mu.Unlock()
	for q.heap.Len() == 0 {
		q.cond.Wait()
	}
	item := heap.Pop(&q.heap).(*entry)
	return item.notification
}

type NotificationDispatcher struct {
	queue *NotificationRequestQueue
}

func NewNotificationDispatcher() *NotificationDispatcher {
	return &NotificationDispatcher{queue: NewNotificationRequestQueue()}
}

func (d *NotificationDispatcher) Enqueue(notification *Notification) {
	d.queue.Offer(notification)
}

func (d *NotificationDispatcher) NextTask() *Notification {
	return d.queue.Take()
}

const maxRetries = 3

type NotificationWorker struct {
	dispatcher        *NotificationDispatcher
	preferenceService *UserPreferenceService
	channelFactory    *NotificationChannelFactory
	deliveryWG        sync.WaitGroup
}

func NewNotificationWorker(dispatcher *NotificationDispatcher, preferenceService *UserPreferenceService, channelFactory *NotificationChannelFactory) *NotificationWorker {
	return &NotificationWorker{
		dispatcher:        dispatcher,
		preferenceService: preferenceService,
		channelFactory:    channelFactory,
	}
}

func (w *NotificationWorker) Run() {
	for {
		notification := w.dispatcher.NextTask()
		if notification == nil { // poison pill
			w.deliveryWG.Wait()
			return
		}

		userPref := w.preferenceService.GetPreferences(notification.RecipientID)
		var channelsToUse []NotificationType
		if len(userPref) == 0 {
			channelsToUse = []NotificationType{notification.Type}
		} else {
			for t := range userPref {
				channelsToUse = append(channelsToUse, t)
			}
		}

		for _, t := range channelsToUse {
			w.deliveryWG.Add(1)
			go func(n *Notification, t NotificationType) {
				defer w.deliveryWG.Done()
				w.deliverWithRetry(n, t)
			}(notification, t)
		}
	}
}

func (w *NotificationWorker) deliverWithRetry(notification *Notification, t NotificationType) {
	channel, err := w.channelFactory.GetNotificationChannel(t)
	if err != nil {
		fmt.Println(err)
		notification.Status = FAILED
		return
	}

	for attempt := 1; attempt <= maxRetries; attempt++ {
		if channel.SendNotification(notification) {
			notification.Status = SENT
			return
		}
	}
	notification.Status = FAILED
	fmt.Printf("Notification %s failed after %d attempts\n", notification.ID, maxRetries)
}

type NotificationService struct {
	dispatcher *NotificationDispatcher
}

func NewNotificationService(dispatcher *NotificationDispatcher) *NotificationService {
	return &NotificationService{dispatcher: dispatcher}
}

func (s *NotificationService) SubmitNotificationRequest(notification *Notification) bool {
	s.dispatcher.Enqueue(notification)
	return true
}

func main() {
	preferenceService := NewUserPreferenceService()
	channelFactory := NewNotificationChannelFactory()
	dispatcher := NewNotificationDispatcher()
	notificationService := NewNotificationService(dispatcher)

	alice := &User{ID: "u1", Name: "Alice"}
	preferenceService.SetPreferences("u1", []NotificationType{EMAIL, SMS})

	worker := NewNotificationWorker(dispatcher, preferenceService, channelFactory)
	var workerWG sync.WaitGroup
	workerWG.Add(1)
	go func() {
		defer workerWG.Done()
		worker.Run()
	}()

	notificationService.SubmitNotificationRequest(&Notification{
		ID: "n1", User: alice, RecipientID: "u1", Message: "Your order has shipped!",
		Priority: 1, Type: EMAIL, Status: PENDING,
	})
	notificationService.SubmitNotificationRequest(&Notification{
		ID: "n2", User: alice, RecipientID: "u1", Message: "OTP: 4821",
		Priority: 0, Type: SMS, Status: PENDING,
	})

	time.Sleep(500 * time.Millisecond)
	dispatcher.Enqueue(nil) // poison pill to stop the worker
	workerWG.Wait()

	fmt.Println("Done.")
}
